lib/pbench/server/api/resources/query_apis/iteration_samples.py

```diff
@@ -0,0 +1,245 @@
+            abort(HTTPStatus.INTERNAL_SERVER_ERROR, message="Dataset owner not found")
+
+        # For Iteration samples, we check authorization against the ownership of the
+        # dataset that was selected rather than having an explicit "user"
+        # JSON parameter. This will raise UnauthorizedAccess on failure.
+        self._check_authorization(owner.username, dataset.access)
+
+        # The dataset exists, and authenticated user has enough access so continue
+        # the operation with the appropriate CONTEXT.
+        return {"dataset": dataset, "run_id": run_id}
+
+
+class IterationSampleNamespaces(IterationSamples):
+    """
+    Iteration samples API that returns only unique namespaces for each
+    whitelisted result-data-sample subdocuments.
+
+    This class inherits the common IterationSamples class and builds an
+    aggregated query against a user supplied run id to compile the filter header
+    which shows every unique field in the 'benchmark', 'sample', 'iterations'
+    and 'run' subdocuments.
+    """
+
+    def __init__(self, config: PbenchServerConfig, logger: Logger):
+        super().__init__(config, logger)
+
+    def get_keyword_fields(self, mappings):
+        fields = []
+
+        def recurse(prefix, mappings, result):
+            if "properties" in mappings:
+                for p, m in mappings["properties"].items():
+                    recurse(f"{prefix}{p}.", m, result)
+            elif mappings.get("type") == "keyword":
+                result.append(prefix[:-1])  # Remove the trailing dot, if any
+            elif mappings.get("type") == "text":
+                for f, v in mappings.get("fields", {}).items():
+                    recurse(f"{prefix}{f}.", v, result)
+
+        recurse("", mappings, fields)
+        return fields
+
+    def assemble(self, json_data: JSON, context: CONTEXT) -> JSON:
+        """
+        Construct a pbench search query for aggregating unique values for keyword
+        fields in the 'benchmark', 'sample', 'iterations' and 'run'
+        sub-documents of result-data-sample index document that belong to the
+        given run id.
+        {
+            "run_id": "run id string"
+        }
+        json_data: JSON dictionary of type-normalized parameters
+            "run_id": String representation of run id
+        """
+        run_id = context.get("run_id")
+        dataset = context.get("dataset")
+
+        self.logger.info(
+            "Return sample iteration namespaces for user {}, prefix {}: for "
+            "run id: {}",
+            Auth.token_auth.current_user().username,
+            self.prefix,
+            run_id,
+        )
+
+        # Retrieve the ES index that belongs to this run_id from the metadata
+        # table
+        try:
+            index_map = Metadata.getvalue(dataset=dataset, key="server.index-map")
+            index_keys = [key for key in index_map if "result-data-sample" in key]
+            assert len(index_keys) == 1
```

Comment:
I don't really like the `assert` in server code, because it'll take down the server process. The way gunicorn works I think our server instances are actually single-threaded, but I'm not sure the gunicorn instance would be restarted unless the root gunicorn process crashes, so I think we've just reduced our parallelism permanently. This should be an `abort` with an internal error and a log to tell us what happened.